CIP-0036/README.md

```diff
@@ -0,0 +1,141 @@
+---
+CIP: 36
+Title: New Plutus built-in serialiseBuiltinData
+Authors: Matthias Benkort <matthias.benkort@iohk.io>, Sebastian Nagel <sebastian.nagel@iohk.io> 
+Discussions-To: https://github.com/cardano-foundation/CIPs/pull/218
+Comments-URI: https://github.com/cardano-foundation/CIPs/pull/218
+Status: Draft
+Type: Standards Track
+Created: 2022-02-09
+License: Apache-2.0
+Requires: CIP-35
+
+---
+
+# New Plutus built-in serialiseBuiltinData
+
+## Abstract
+
+This document describes the addition of a new Plutus builtin for serialising `BuiltinData` to `BuiltinByteString`.
+
+## Motivation
+
+As part of developing on-chain script validators for [the Hydra Head protocol](https://eprint.iacr.org/2020/299), we stumble across a peculiar need for on-chain scripts: we need to verify and compare digests obtained from hashing elements of the script's surrounding transaction. 
+
+In this particular context, those elements are transaction outputs (a.k.a. `TxOut`). While Plutus already provides built-in for hashing data-structure (e.g. `sha2_256 :: BuiltinByteString -> BuiltinByteString`), it does not provide generic ways of serialising some data type to `BuiltinByteString`.
+
+In an attempt to pursue our work, we have implemented [an on-chain library (plutus-cbor)][plutus-cbor] for encoding data-types as structured [CBOR / RFC 8949][CBOR] in a _relatively efficient_ way (although still quadratic, it is as efficient as it can be with Plutus' available built-ins) and measured the memory and CPU cost of encoding `TxOut` **in a script validator on-chain**. 
```

Comment:
I'm confused: here you say that your implementation is quadratic, but below you say it's linear (and the graph agrees with that).

Reply:
We would expect a quadratic growth, but only see a linear. Maybe due to this: https://input-output-rnd.slack.com/archives/C21UF2WVC/p1644899404438899?thread_ts=1644875367.010429&cid=C21UF2WVC?

Reply:
^ I think that's unlikely to account for this.  Costs of bytestring operations are based on the size of the bytestring, and that was a change that changed the size of the empty bytestring from 1 to 0 and hence caused slight cost differences between the cost of comparing bytestrings in two different Plutus Core versions.  It shouldn't affect the basic shape of the cost function for appending bytestrings, which is linear in the sum of the sizes of the arguments.

Reply:
I think these graphs are in fact parabolas, but you haven't gone far enough out.
I increased the maximum list size in `Main.hs` to 500 and changed `pparams` in Vaildator.hs to ` def{_maxTxExUnits = ExUnits 9999999999999999 9999999999999999}` and I got the following graph (blue=memory, green=cpu):

![Serialisation](https://user-images.githubusercontent.com/2124565/155064628-424b2a0e-e90a-409f-b1cd-561c658c8253.png)

I think the explanation is that bytestring concatenation is pretty cheap.  The current cpu cost of calling our `appendBytestring` function with arguments of size `a` and `b` is given by `396231 + 621*(a+b)`, which increases very slowly with `a` and `b` (sizes here are unfortunately measured in 64-bit words, not bytes, and the literal numbers are notionally picoseconds).   If you add up a lot of these, `a` and `b` have to get pretty big before the `a+b` term becomes significant relative to the constant term, so this makes the total cost of adding up lots of bytestrings look linear for small inputs.  I couldn't see how to work out the sizes of the bytestrings involved in your examples,  but if we knew that then we could check that they're not large enough to cause quadratic behaviour to become apparent.

The memory cost is just given by `a+b` (ie, the size of the result) and again is measured in 64-bit words, so I think you'd have to be adding up quite a large number of large bytestrings to see quadratic memory usage; again, knowing the sizes of the things being concatenated would be useful.

We got the numbers for the cpu cost function by using Criterion to run the `apppendBytestring` function with inputs of sizes up to 5000, and then fitting a linear function to the execution times; the model fits the data pretty closely, so I think that concatenation times do increase pretty slowly, presumably because the [underlying function](https://hackage.haskell.org/package/bytestring-0.11.2.0/docs/src/Data.ByteString.Internal.html#concat) in `Data.ByteString` is calling C's ``memcpy` to do the hard work.